### ISSUE TITLE
Output shape does not reflect dilation_patch parameter.

### ISSUE THREAD

**Issue by nehpetsde** [ClementPinard/Pytorch-Correlation-extension#35]
The `dilation_patch` parameter seems not correctly interpreted:
```python
>>> import torch
>>> from spatial_correlation_sampler import SpatialCorrelationSampler
>>> x = torch.ones(1, 1, 10, 10)
>>> SpatialCorrelationSampler(patch_size=4, dilation_patch=1)(x, x).shape
torch.Size([1, 5, 5, 10, 10])
>>> SpatialCorrelationSampler(patch_size=5, dilation_patch=2)(x, x).shape
torch.Size([1, 5, 5, 10, 10])
>>> SpatialCorrelationSampler(patch_size=5, dilation_patch=3)(x, x).shape
torch.Size([1, 5, 5, 10, 10])
```
Regardless of the stepping within the patch the output shape is just the same.

**Comment by 0point618** [comment]
I also encountered the same problem

**Comment by ClementPinard** [comment]
My results :
```
In [8]: SpatialCorrelationSampler(patch_size=4, dilation_patch=1)(x, x).shape                                                                                                                              
Out[8]: torch.Size([1, 4, 4, 10, 10])

In [9]: SpatialCorrelationSampler(patch_size=5, dilation_patch=2)(x, x).shape                                                                                                                              
Out[9]: torch.Size([1, 5, 5, 10, 10])

In [10]: SpatialCorrelationSampler(patch_size=5, dilation_patch=3)(x, x).shape                                                                                                                             
Out[10]: torch.Size([1, 5, 5, 10, 10])

```
Not sure why you get `[1,5,5,10,10]` for the first result (hope it got fixed when i updated the package), but the other results are the right ones.

as per paragraph in README, output shape is `[B x PatchH x PatchW x oH x oW]`, so the first 2 dimensions after the batch dimension are expected to be patch size. dilation_patch does not change that, the second feature map is autmatically padded to fit (or rather we only compute partial patches at the edge, see [here](https://github.com/ClementPinard/Pytorch-Correlation-extension/blob/master/Correlation_Module/correlation.cpp#L23) ).

Moreover, output size `oH`and `oW` are only dependant on the actual correlation operation, which is defined by its kernel and its stride. Should you have a kernel of size 3 like most convolutions, you would have `oH = oW = 8`

A better to understand what is done and what output should be expected is to look at the algorithm from another (less efficient) point of view :
 - shift the 2nd feature map with a particular vector (u,v) (u is within [-patchH/2 * dilation_patch, patchH/2 * dilation_patchH], same for v with W)
 - computed the correlation between 1st feature map and shifted second feature map with a particular kernel. For a kernel of size 1 (as in FlowNetC), this is just a term wise multiplication.

As such, you can see that 1st feature map is NEVER shifted, and the operation is NOT symmetrical. Also, should we have a fixed padding for shifted 2nd feature map, we would have to make a stack of layers that are not the same size which is not possible. We could crop everything to the lowest size so that we never hit bounds, but that would mean that the correlation with a 0 shift would be incomplete, which is unacceptable IMO.

See a discussion on the pytorch github about different versions of the same algorithm : https://github.com/pytorch/pytorch/issues/4073
